### PR TITLE
Add MacOS COREML support for vsort

### PR DESCRIPTION
**Pull request by yuygfgg**

### CODE REVIEW

**Comment by WolframRhodium** [comment]
I do not have an Apple device, could you please check if Core ML is working?

**Comment by yuygfgg** [comment]
> I do not have an Apple device, could you please check if Core ML is working? Thanks in advance.

sadly not.

**Comment by WolframRhodium** [comment]
DML is Windows only.

**Comment by yuygfgg** [comment]
I fixed the libc++abi error. It's not vs-mlrt issue but homebrew's. Replacing with github release fixed the problem

**Comment by WolframRhodium** [comment]
Great! What models are you testing? Is the output correct, and what's the performance?

**Comment by yuygfgg** [comment]
> Great! What models are you testing? Is the output correct, and what's the performance? (it should be faster than [MTL](https://github.com/AmusementClub/vs-mlrt/wiki/Intel-Core-Ultra-7-155H)

I haven't done a full test yet. I saw you are using ORT_SEQUENTIAL instead of ORT_PARALLEL? Is that intended? Because as you can see in the screenshot, ANE isn't at 100%

**Comment by yuygfgg** [comment]
> It's deliberate. ORT_PARALLEL is only used for CPU parallel execution, and may cause problems on accelerators (from my experience).
Ok I will do more research to optimize the performance. I'm only getting <1 fps on dpir gray at the moment.

**Comment by WolframRhodium** [comment]
You can also compare the performance of PyTorch's MPS backend of dpir. It's likely that coreml is not optimized.

**Comment by yuygfgg** [comment]
> You can also compare the performance of PyTorch's MPS backend for dpir. It's likely that coreml is not optimized.

mps is running on gpu instead of ane. And it also gives me less than 1fps but used 5 times power.

**Comment by yuygfgg** [comment]
I'm not able to use fp16 on COREML for some reason.

It runs only on CPU if I set FP16 True.

**Comment by WolframRhodium** [comment]
[This](https://github.com/hollance/neural-engine) document looks interesting. I will provide feedback after reading it.

**Comment by yuygfgg** [comment]
> According to [this](https://github.com/hollance/neural-engine/blob/master/docs/16-bit.md#is-the-ane-16-bit), fp16 should be supported, but the framework may decide to run on CPU only. (And MPS in PyTorch is GPU only, that makes sense.)
> 
> What about `ArtCNN` models? Those models are pretty standard and should be well optimized.

ArtCNN_R16F96 runs at 0.7 fps and ArtCNN_R8F64 2-3 fps. Both with 85% ANE usage

**Comment by WolframRhodium** [comment]
Is fp16 enabled?

> Both with 85% ANE usage

That's great.

**Comment by yuygfgg** [comment]
> ArtCNN_R8F64

FP16 is not supported by onnx on COREML, according to https://github.com/microsoft/onnxruntime/issues/17448

**Comment by yuygfgg** [comment]
> Is fp16 enabled?
> 
> > Both with 85% ANE usage
> 
> That's great.

FP16 is not supported by onnx on COREML, according to https://github.com/microsoft/onnxruntime/issues/17448

**Comment by WolframRhodium** [comment]
I see.

**Comment by WolframRhodium** [comment]
I prefer to use `ORT_SEQUENTIAL` and `num_streams=1` by default.

**Comment by WolframRhodium** [comment]
Thanks, LGTM.

File: vsort/vs_onnxruntime.cpp
Comment (yuygfgg) [inline vsort/vs_onnxruntime.cpp:1]: The extra "else" here caused compile failure. Will it also be a problem for DML?

Comment (WolframRhodium) [reply vsort/vs_onnxruntime.cpp:1]: I can fix it later, thanks.

**Pull request merged**